src/style-spec/function/expression.js

```diff
@@ -168,6 +198,103 @@ class LambdaExpression extends BaseExpression {
     }
 }
 
+class Reference extends BaseExpression {
+    name: string;
+    constructor(key: string, name: string, type: Type) {
+        super(key, type);
+        if (!/^[a-zA-Z_]+[a-zA-Z_0-9]*$/.test(name))
+            throw new ParsingError(key, `Invalid identifier ${name}.`);
```

Comment:
This check should be moved to `LetExpression.parse()`, and it should also either check against JS reserved words or else 'escape' them (e.g. replace `function` with `var_function` or something)